### ISSUE TITLE
Can not login with UPN.

### ISSUE THREAD

**Issue by svhsvh** [unosquare/passcore#220]
**PassCore Server**
- OS: [Windows ]
- Provider: [Active Directory ]

**Describe the bug**
Our users are used to login with their UPN, but they are not able to login with their UPN. If they try passcore returns: "Please enter a valid username", like you can see in the image below:

**Screenshots**
![image](https://user-images.githubusercontent.com/43951839/46599490-c53a7980-cae7-11e8-8460-01370b3c0b25.png)

I hope someone knows the solution to our problem or is able to help us, thanks in advance.

**Comment by diogenes25** [comment]
The identity is searched by default over the "cn". If you want to find the identity through the UPN, you can change line 26 in File [PasswordChangeProvider](https://github.com/unosquare/passcore/blob/dccaa55c9b13e92a5197d730bdb4315e48ad0eb8/src/Unosquare.PassCore.PasswordProvider/PasswordChangeProvider.cs#L26).cs  as followed.

```csharp
 public ApiErrorItem PerformPasswordChange(string username, string currentPassword, string newPassword)
        {
            // perform the password change
            try
            {
                using (var principalContext = AcquirePrincipalContext())
                {
                   // ########## Replace this line ... ##################
                   //  var userPrincipal = UserPrincipal.FindByIdentity(principalContext, username);
                   //  ####### with this line !!! Set ItenditityType to: IdentityType.UserPrincipalName #####
                    var userPrincipal = UserPrincipal.FindByIdentity(principalContext, IdentityType.UserPrincipalName, username);

            .....

```

**Comment by svhsvh** [comment]
Thanks, where is that file located? because I'm unable to locate it.

**Comment by diogenes25** [comment]
Here:  [passcore/src/Unosquare.PassCore.PasswordProvider/PasswordChangeProvider.cs Line: 26](https://github.com/unosquare/passcore/blob/dccaa55c9b13e92a5197d730bdb4315e48ad0eb8/src/Unosquare.PassCore.PasswordProvider/PasswordChangeProvider.cs#L26)

**Comment by svhsvh** [comment]
But that is the source code right? Because that file is not present on my passcore installation.

**Comment by diogenes25** [comment]
Yes it is.
You must change it the "hard way" (with compile and all the stuff)
We had a similar Problem here  #204.
Maybe we should put a switch in appsettings.json that defines how a user is found (Distinguished Name (DN), Globally Unique Identifier (GUID), Security Account Manager (SAM), Security Identifier (SID) or User Principal Name (UPN)).

**Comment by geoperez** [comment]
I like your idea.

**Comment by svhsvh** [comment]
Thanks for your answer. Yes that would be nice!

**Comment by geoperez** [comment]
@svhsvh you can download the current master code or wait for the release probably by Friday.

**Comment by svhsvh** [comment]
Thanks! I will wait until the release.

**Comment by svhsvh** [comment]
Just a quick question, where can I find it when this will be released?

**Comment by geoperez** [comment]
Sorry guys, I've been busy but probably today I'll release it.

**Comment by geoperez** [comment]
@svhsvh I created a published version from my current branch. Including:

* The PR to define DC Attrbibute by @diogenes25 
* New appsettings format
* Logging improved.

[PassCore35.zip](https://github.com/unosquare/passcore/files/2488558/PassCore35.zip)

Let me know if it's working for you. I'll wait feedback before to publish a release.

**Comment by MAP74** [comment]
This version works with UPN.
But it fails when user has to change password at next logon.
Error message: You are not allowed to change your password. Please contact your system administrator.

The only version if found where this works ouf of the box is v3.1

**Comment by geoperez** [comment]
Are you sure about that? Because the Change Password Provider from [3.1 ](https://github.com/unosquare/passcore/blob/3.1.0/src/Unosquare.PassCore.Web/Helpers/PasswordChangeProvider.cs) is pretty much the same code as this version. The only important change is the ability to change the DC attribute to search the user. 

There are no changes in how the password is changed.

**Comment by svhsvh** [comment]
I'am not able to get PassCore35 working, I get the HTTP Error 502.5. I do not understand why because the stable release is working perfect and on the same server. And I have .NET Core 2.1.1 Windows Server Hosting bundle installed.
<img width="643" alt="screen shot 2018-10-29 at 10 07 02" src="https://user-images.githubusercontent.com/43951839/47640047-006a1e80-db63-11e8-9812-9b8278ce2dee.png">

**Comment by geoperez** [comment]
Do you have a log file?

**Comment by svhsvh** [comment]
Unfortunately not.

**Comment by svhsvh** [comment]
Or has PassCore35 more dependencies?

**Comment by MAP74** [comment]
> 
> 
> Are you sure about that? Because the Change Password Provider from [3.1 ](https://github.com/unosquare/passcore/blob/3.1.0/src/Unosquare.PassCore.Web/Helpers/PasswordChangeProvider.cs) is pretty much the same code as this version. The only important change is the ability to change the DC attribute to search the user.
> 
> There are no changes in how the password is changed.

Hi,

I did some test to be really sure.

Results below.

Thread that triggered me to try 3.1: https://github.com/unosquare/passcore/issues/216

Test setup:
-----------
AD functional level: 2016
Passcore runs on:
OS: Windows Server 2016 (Member)
Provider: Active Directory
IIS: v10

Password-Policy:
-------------------
Complexity:   Enabled
History:	     Enabled 
Max-Age:	     Enabled
Min-Age:	     Enabled 
Min-Length:  Enabled

Test-User:
----------
Default domain user with "User must change password at next logon" enabled.

######################
**Passcore versions tested:**
######################
-----------------------------


3.1.0
-----
	App-Pool IDs:	ApplicationPoolIdentity, Custom AppPool User with AD rights granted, NetworkService
	Message-WebIf:	You have changed your password successfully.
			Please note it may take a few hours for your new password to reach all domain controllers.
	Message-Logfi:	no usable error messages found
					(Loglevel:Debug)

3.2.0 (Upgrade to new ASP.NET Core 2.1)
-----
	App-Pool IDs:	ApplicationPoolIdentity, Custom AppPool User with AD rights granted, NetworkService
	Message-WebIf:	You have changed your password successfully.
			Please note it may take a few hours for your new password to reach all domain controllers.
	Message-Logfi:	no usable error messages found
					(Loglevel:Debug)

3.3.0
-----
	App-Pool IDs:	ApplicationPoolIdentity, Custom AppPool User with AD rights granted, NetworkService
	Message-WebIf:	Access is denied.
	Message-Logfi:	no usable error messages found
					(Loglevel:Debug)

3.4.2
-----
	App-Pool IDs:	ApplicationPoolIdentity, Custom AppPool User with AD rights granted, NetworkService
	Message-WebIf:	Access is denied.
	Message-Logfi:	no usable error messages found
					(Loglevel:Debug)

3.5.0
-----
	App-Pool IDs:	ApplicationPoolIdentity, Custom AppPool User with AD rights granted, NetworkService
	Message-WebIf:	Failed to update password: Access is denied.
	Message-Logfi:	info: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  PerformPasswordChange for user mpxxx@xyz.com
			info: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  PerformPasswordChange for user mpxxx@xyz.com
			warn: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  Using AutomaticContext
			warn: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  Using AutomaticContext	
			warn: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  The User principal password have no last password
			warn: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  The User principal password have no last password
					(Loglevel:Debug)

					
					
Changed default settings in all versions for this test:
-------------------------------------------------------
```json
appsettings.json
#########################################
"Logging": {
	"LogLevel": {
		"Default":   "Debug",
		"System":    "Debug",
		"Microsoft": "Debug" }
}

"AppSettings": {"DefaultDomain": "FQDN of our AD domain"}
	and
"AppSettings": {"DefaultDomain": ""}

web.config
#########################################
stdoutLogEnabled="true"
```


Current setup running without problems:
---------------------------------------
Right now I'm on v3.2 with AppPoolID "Network Service".

**Comment by geoperez** [comment]
@svhsvh has a different setting file, did you check that?

**Comment by geoperez** [comment]
@MAP74 wow! thank you for the information. Let me review again the changes between 320 and 330.

**Comment by geoperez** [comment]
I noticed a changed introduce after 3.2.0 to set the last password property, and I guess it may could affect. The following release includes a new setting named `UpdateLastPassword` (default value false). If the value is true will execute the last password check, otherwise, continue to password changing.

[PassCore35.zip](https://github.com/unosquare/passcore/files/2530566/PassCore35.zip)

For more information about the Last Password check, see issue #21.

**Comment by MAP74** [comment]
@geoperez The new 3.5 you provided does the trick. Changing passwords with option "User must change password at next logon" enabled is working again,

Log
---
	Message-Logfi:	info: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  PerformPasswordChange for user mpx@xyz.com
			info: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  PerformPasswordChange for user mpx@xyz.com
			warn: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  Using AutomaticContext
			warn: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  Using AutomaticContext
			dbug: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  The User principal password updated with setPassword
			dbug: Unosquare.PassCore.PasswordProvider.PasswordChangeProvider[0]
				  The User principal password updated with setPassword

Thank you.

Question
---
The PasswordMeter seems to use the classic complexity requirements to measure if your password is a strong one.

```Passw0rd%```
is considered a lot stronger than
```this is an absolute insane long password if you have to type it in every time but easy to remember```
Ironically it provides a link to [https://xkpasswd.net/s/](url) where we are teached why classic passwords are unsafe and why long and easy to remember passwords with a high entropy should be preferred.

Is this something where the password meter can be improved?

**Comment by geoperez** [comment]
That's cool! Regarding the PasswordMeter, you are right, the current implementation is not right. But I'll target that issue in the following version. I want to release 3.5.0 now to fix the issues between 3.2.0 and 3.4.0.

Can you submit this like a new issue?
